pkg/credentialprovider/external.go

```diff
@@ -0,0 +1,232 @@
+				value = "https://" + value
+			}
+			parsed, err := url.Parse(value)
+			if err != nil {
+				klog.Errorf("Entry %q in dockercfg invalid (%v), ignoring", m, err)
+				continue
+			}
+
+			// The docker client allows exact matches:
+			//    foo.bar.com/namespace
+			// Or hostname matches:
+			//    foo.bar.com
+			// It also considers /v2/  and /v1/ equivalent to the hostname
+			// See ResolveAuthConfig in docker/registry/auth.go.
+			effectivePath := parsed.Path
+			if strings.HasPrefix(effectivePath, "/v2/") || strings.HasPrefix(effectivePath, "/v1/") {
+				effectivePath = effectivePath[3:]
+			}
+			var key string
+			if (len(effectivePath) > 0) && (effectivePath != "/") {
+				key = parsed.Host + effectivePath
+			} else {
+				key = parsed.Host
+			}
+
+			keyring.providers[key] = &provider
+			keyring.index = append(keyring.index, key)
+		}
+	}
+
+	eliminateDupes := sets.NewString(keyring.index...)
+	keyring.index = eliminateDupes.List()
+
+	// Update the index used to identify which credentials to use for a given
+	// image. The index is reverse-sorted so more specific paths are matched
+	// first. For example, if for the given image "quay.io/coreos/etcd",
+	// credentials for "quay.io/coreos" should match before "quay.io".
+	sort.Sort(sort.Reverse(sort.StringSlice(keyring.index)))
+
+	return keyring, nil
+}
+
+func (e *externalProviderKeyring) Lookup(image string) ([]AuthConfig, bool) {
+	var provider ExternalCredentialProvider
+	klog.V(2).Infof("Looking up %s with external provider.", image)
+
+	// range over the index as iterating over a map does not provide a predictable ordering
+	for _, matcher := range e.index {
+		// both k and image are schemeless URLs because even though schemes are allowed
+		// in the credential configurations, we remove them in Add.
+		if matched, _ := urlsMatchStr(matcher, image); matched {
+			provider = e.providers[matcher]
+			break
+		}
+	}
+
+	if provider == nil {
+		klog.Infof("No external provider available for the image: %s", image)
+		return []AuthConfig{}, false
+	}
+
+	cfg := provider.Provide(image)
+
+	providedConfigs := make([]AuthConfig, 0)
+	for _, ident := range cfg {
+		creds := AuthConfig{
+			Username: ident.Username,
+			Password: ident.Password,
+			Email:    ident.Email,
+		}
+		providedConfigs = append(providedConfigs, creds)
+	}
+	return providedConfigs, (len(providedConfigs) > 0)
+}
+
+type ExternalCredentialProvider interface {
+	Provide(image string) DockerConfig
+}
+
+type externalCredentialProvider struct {
+	command string
+	args    []string
+
+	// Each environment variable is in the form key=value
+	env []string
+}
+
+func (e *externalCredentialProvider) Provide(image string) DockerConfig {
+	cfg := make(DockerConfig)
+
+	response, err := e.request(image)
+	if err != nil {
+		klog.Errorf("Failed getting credential from external registry credential provider: %v", err)
+		return cfg
+	}
+
+	cfg[image] = DockerConfigEntry{
```

Comment:
I do not understand the point of this map based structure (which is later transformed to a list).  Seems like we are only returning one item.  Are we trying to maintain an old interface which does not apply to this code?